### PR TITLE
Tag group form

### PR DESCRIPTION
**Pull request by brunohkbx**
tag_groups

tag_groups2

### CODE REVIEW

File: app/views/projects/_form.html.erb [hunk 1]
```diff
@@ -32,6 +32,11 @@
   </div>
 
   <div class="form-group">
+    <%= f.label :tag_group, class: 'control-label' %>
+    <%= f.select :tag_group_id, tag_group_options, {:include_blank => true}, { class: 'form-control auth-form-control' } %>
```
Comment (andrerpbts) [inline app/views/projects/_form.html.erb]: The `+` button, to open a modal with a new tag group form here, like we talked before, is missing.

File: app/views/tag_groups/_form.html.erb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  <div class="form-group">
+    <%= f.label :name, class: 'control-label' %>
+    <%= f.text_field :name, class: 'form-control auth-form-control' %>
+  </div>
+
+  <div class="form-group">
+    <%= f.label :description, class: 'control-label' %>
+    <%= f.text_area :description, class: 'form-control auth-form-control' %>
+  </div>
+  <div class="form-group">
```
Comment (andrerpbts) [inline app/views/tag_groups/_form.html.erb]: Add a space between this divs...

File: app/views/tag_groups/edit.html.erb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    <div class="page-header">
+      <h4 class="page-header-title darkgrey-2">
+        <%= t('projects.edit tag_group') %>
+      </h4>
+    </div>
+  </div>
+  <div class="col-xs-12 col-sm-5 col-sm-offset-2">
+    <div class="panel card">
+      <div class="panel-body"
+      data-current_color=<%= @tag_group[:bg_color] %> >
```
Comment (andrerpbts) [inline app/views/tag_groups/edit.html.erb]: align this with class attribute. 

Optionally you can rename this to `data-current-color`.

File: app/assets/javascripts/components/tag_groups/ColorPick.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    };
+
+    this.handleChange = (color) => {
+      this.setState({ color: color.hex })
+    };
+  }
+
+  render() {
+
+    const styles = reactCSS({
```
Comment (andrerpbts) [inline app/assets/javascripts/components/tag_groups/ColorPick.js]: I think you could move those styles to a proper `.scss` file.

File: app/assets/javascripts/controllers/tag_groups/IndexController.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import User from 'models/user';
+import ColorPick from 'components/tag_groups/ColorPick';
+import ProjectCollection from 'collections/project_collection';
+
+export default () => {
+  const bgColor = $('.panel-body').data('current_color') || "#F17013";
+
+  ReactDOM.render(<ColorPick color = { bgColor }/>, document.getElementById('tag_group_color'));
```
Comment (andrerpbts) [inline app/assets/javascripts/controllers/tag_groups/IndexController.js]: Like we do in normal HTML attributes, you can write this without spaces between attribute, equal sign and value, like:

```javascript
ReactDOM.render(<ColorPick color={ bgColor } />, document.getElementById('tag_group_color'));
```

File: app/controllers/projects_controller.rb [hunk 6]
```diff
@@ -233,7 +233,7 @@ def set_story_flow
   end
 
   def allowed_params
-    params.fetch(:project,{}).permit(:name, :point_scale, :default_velocity, :start_date, :iteration_start_day, :iteration_length, :import, :archived, :disallow_join)
+    params.fetch(:project,{}).permit(:name, :point_scale, :default_velocity, :tag_group_id, :start_date, :iteration_start_day, :iteration_length, :import, :archived, :disallow_join)
```
Comment (andrerpbts) [inline app/controllers/projects_controller.rb]: Please, break this in multiple lines. Line too long here.

Also, use `require(:project)` instead `fetch`.

File: app/controllers/tag_groups_controller.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+  # POST /tag_groups/new
+  def create
+    @tag_group = current_team.tag_groups.new(allowed_params)
+    authorize @tag_group
+
+    if @tag_group.save
+      redirect_to tag_groups_path
+    else
+      render :edit
```
Comment (andrerpbts) [inline app/controllers/tag_groups_controller.rb]: I think you did mean `render :new` here, no?

File: app/controllers/tag_groups_controller.rb [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+  def destroy
+    @tag_group = current_team.tag_groups.find(params[:id])
+    authorize @tag_group
+    @tag_group.destroy
+    redirect_to tag_groups_path
+  end
+
+  def allowed_params
+    allowed_params = params.require(:tag_group).permit(:name, :description, :bg_color)
```
Comment (andrerpbts) [inline app/controllers/tag_groups_controller.rb]: this variable association isn't necessary here.

File: app/models/tag_group.rb [hunk 9]
```diff
@@ -3,4 +3,16 @@ class TagGroup < ActiveRecord::Base
   belongs_to :team
 
   validates :name, presence: true, length: { maximum: 15 }
+
+  def bg_color=(value)
```
Comment (andrerpbts) [inline app/models/tag_group.rb]: Missing test for this method.

Don't forget to test an invalid JSON too, because its behaviour.

File: app/models/tag_group.rb [hunk 10]
```diff
@@ -3,4 +3,16 @@ class TagGroup < ActiveRecord::Base
...
+    write_attribute(:bg_color, convert_bg_color_hash(value))
+  end
+
+  private
+
+  def convert_bg_color_hash(value)
+    RGBUtils::RGB.new(JSON.parse(value)).to_hex
```
Comment (andrerpbts) [inline app/models/tag_group.rb]: I think you can write this inline:

```ruby
def convert_bg_color_hash(value)
  RGBUtils::RGB.new(JSON.parse(value)).to_hex rescue value
end
```

File: app/policies/tag_group_policy.rb [hunk 11]
```diff
@@ -0,0 +1,3 @@
+class TagGroupPolicy < StoryPolicy
+
```
Comment (andrerpbts) [inline app/policies/tag_group_policy.rb]: Need to think some policies methods here...

For me it's simple, if user is admin, then he can handle all tag groups.

File: app/views/tag_groups/edit.html.erb [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+<div class="row">
+  <div class="col-xs-12 col-sm-5 col-sm-offset-2">
+    <div class="page-header">
+      <h4 class="page-header-title darkgrey-2">
+        <%= t('projects.edit tag_group') %>
+      </h4>
+    </div>
+  </div>
+  <div class="col-xs-12 col-sm-5 col-sm-offset-2">
```
Comment (andrerpbts) [inline app/views/tag_groups/edit.html.erb]: Space each div block...

File: app/views/tag_groups/index.html.erb [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+                    <div class="color_container">
+                      <div class="color" style="background-color: <%= tag_group.bg_color %>;">
+                      </div>
+                    </div>
+                  </td>
+                  <td><%= tag_group.name %></td>
+                  <td><%= tag_group.description %></td>
+                  <td>
+                    <%= link_to 'Edit', edit_tag_group_path(tag_group) %>
+                    <%= link_to "Delete", tag_group, :method => :delete, :data => {:confirm => 'Are you sure?'} %>
```
Comment (andrerpbts) [inline app/views/tag_groups/index.html.erb]: Use single quotes here, like you did in Edit.

Also, can you style this in a "danger" color? 💃 💅

File: app/views/tag_groups/index.html.erb [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+                  </td>
+                </tr>
+              <% end -%>
```
Comment (andrerpbts) [inline app/views/tag_groups/index.html.erb]: this minus isn't neccessary here.

File: app/views/tag_groups/new.html.erb [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+
+<div class="row">
+  <div class="col-xs-12 col-sm-5 col-sm-offset-2">
+    <div class="page-header">
+      <h4 class="page-header-title darkgrey-2">
+        <%= t('projects.new tag_group') %>
+      </h4>
+    </div>
+  </div>
+  <div class="col-xs-12 col-sm-5 col-sm-offset-2">
```
Comment (andrerpbts) [inline app/views/tag_groups/new.html.erb]: space div blocks

**Comment by andrerpbts** [comment]
Don't forget the CHANGELOG.

File: app/views/projects/_form.html.erb (hunk 1 above)
Comment (andrerpbts) [inline app/views/projects/_form.html.erb]: Use new hash format in include_blank.

File: app/views/tag_groups/edit.html.erb (hunk 3 above)
Comment (talyssonoc) [inline app/views/tag_groups/edit.html.erb]: I think `data-current-color` is better too, and more compliant with HTML standards. Also, don't you have to wrap it in quotes? `data-current-color="<%= @tag_group[:bg_color] %>`

File: app/assets/javascripts/components/tag_groups/ColorPick.js [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+        },
+      },
+    });
+    return (
+      <div>
+        <div style={ styles.swatch } onClick={ this.handleClick }>
+          <div style={ styles.color } />
+          <input type="hidden" value={ this.state.color } name="tag_group[bg_color]" />
+        </div>
+        { this.state.displayColorPicker ? <div style={ styles.popover }>
```
Comment (talyssonoc) [inline app/assets/javascripts/components/tag_groups/ColorPick.js]: Prefer `{ conditional && something` for JSX instead of `{ conditional ? something : null` where there's not an "else".

File: app/assets/stylesheets/_tag_group.scss [hunk 17]
```diff
@@ -0,0 +1,17 @@
+.tag-groups {
+  margin: 20px 0;
+
+  .color_container {
+    padding: 2px;
+    background: rgb(255, 255, 255);
+    border-radius: 1px;
+    box-shadow: rgba(0, 0, 0, 0.0980392) 0px 0px 0px 1px;
```
Comment (talyssonoc) [inline app/assets/stylesheets/_tag_group.scss]: You could add a variable for this color, seems too specific.

File: app/assets/stylesheets/_tag_group.scss [hunk 18]
```diff
@@ -0,0 +1,17 @@
+.tag-groups {
+  margin: 20px 0;
+
+  .color_container {
+    padding: 2px;
+    background: rgb(255, 255, 255);
```
Comment (talyssonoc) [inline app/assets/stylesheets/_tag_group.scss]: Also use a variable here, I think there's a `$white` already.

File: app/controllers/tag_groups_controller.rb [hunk 19]
```diff
@@ -0,0 +1,58 @@
+class TagGroupsController < ApplicationController
+  skip_after_filter :verify_policy_scoped, only: :index
+
+  # GET /tag_groups
```
Comment (talyssonoc) [inline app/controllers/tag_groups_controller.rb]: O think those comments are not necessary.

File: app/views/tag_groups/_form.html.erb [hunk 20]
```diff
@@ -0,0 +1,21 @@
+<%= form_for(@tag_group) do |f| %>
```
Comment (talyssonoc) [inline app/views/tag_groups/_form.html.erb]: Don't use instance variables inside of partials, pass it explicitly like: `<%= render 'form', tag_groups: @tag_groups %>` and use it with just `tag_groups` here.

File: app/controllers/tag_groups_controller.rb [hunk 21]
```diff
@@ -0,0 +1,58 @@
+class TagGroupsController < ApplicationController
+  skip_after_filter :verify_policy_scoped, only: :index
```
Comment (eddiecode) [inline app/controllers/tag_groups_controller.rb]: I think this method was deprecated in Rails 2.3.8+, and we're using 4.2.7.1 https://apidock.com/rails/ActionController/Filters/ClassMethods/skip_after_filter

File: app/assets/stylesheets/application.scss [hunk 22]
```diff
@@ -32,3 +32,4 @@
 @import "project-reports";
 @import "project-list";
 @import "search-bar";
+@import "tag_group";
```
Comment (eddiecode) [inline app/assets/stylesheets/application.scss]: I think you could use a `-` here instead of `_`, to keep a pattern.

File: app/views/tag_groups/index.html.erb (hunk 13 above)
Comment (eddiecode) [inline app/views/tag_groups/index.html.erb]: Try to use the new hash format, and you need to put the 'Are you sure?' into `I18n` to translate this phrase to others languages.

File: app/assets/stylesheets/_tag_group.scss (hunk 17 above)
Comment (brunohkbx) [inline app/assets/stylesheets/_tag_group.scss]: Can you think of a good name?
It's only differs in opacity.

Comment (talyssonoc) [inline app/assets/stylesheets/_tag_group.scss]: What do you think about `$color-container-shadow-color`? It'll be easier to search for it later.

**Comment by brunohkbx** [comment]
@andrerpbts could you check if the change that I wrote in CHANGELOG is good?

File: app/assets/stylesheets/_tag_group.scss [hunk 23]
```diff
@@ -0,0 +1,17 @@
+.tag-groups {
+  margin: 20px 0;
+
+  .color_container {
+    padding: 2px;
+    background-color: $white;
+    border-radius: 1px;
+    box-shadow: rgba(0, 0, 0, 0.0980392) 0px 0px 0px 1px;
```
Comment (marclerodrigues) [inline app/assets/stylesheets/_tag_group.scss]: You don't have to add units for zero values. `0 0 0 1px`.

File: app/views/tag_groups/_form.html.erb [hunk 24]
```diff
@@ -0,0 +1,22 @@
...
+    <%= f.text_area :description, class: 'form-control auth-form-control' %>
+  </div>
+
+  <div class="form-group">
+    <%= f.label :bg_color, class: 'control-label' %>
+    <div id='tag_group_color'></div>
+  </div>
+
+  <div class="btn-group actions pull-right clearfix">
+    <%= f.submit nil, class: 'btn btn-primary btn-form' %>
```
Comment (marclerodrigues) [inline app/views/tag_groups/_form.html.erb]: Why you're not passing a text to your submit button?

File: app/controllers/tag_groups_controller.rb (hunk 21 above)
Comment (brunohkbx) [inline app/controllers/tag_groups_controller.rb]: What method should I use instead of skip_after_filter ?

Comment (fladson) [inline app/controllers/tag_groups_controller.rb]: `skip_filter` or `skip_after_action`

File: app/assets/javascripts/app.js [hunk 25]
```diff
@@ -1,10 +1,13 @@
 import './libs';
 import ProjectsIndexController from './controllers/projects/IndexController';
+import TagGroupsIndexController from './controllers/tag_groups/IndexController';
 import { start as ProjectsShowController } from './central'
 
 const routes = {
   'projects.index': ProjectsIndexController,
-  'projects.show': ProjectsShowController
+  'projects.show': ProjectsShowController,
+  'tag_groups.new': TagGroupsIndexController,
```
Comment (talyssonoc) [inline app/assets/javascripts/app.js]: Why is it called `TagGroupsIndexController` if it's being used for both `new` and `edit`?

File: app/assets/javascripts/controllers/tag_groups/IndexController.js [hunk 26]
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import User from 'models/user';
```
Comment (talyssonoc) [inline app/assets/javascripts/controllers/tag_groups/IndexController.js]: Why are we importing `User` and `ProjectCollection` but not using it?

File: app/assets/stylesheets/_color-picker.scss [hunk 27]
```diff
@@ -0,0 +1,27 @@
+.swatch {
+  padding: 5px;
+  background: $white;
+  border-radius: 1px;
+  box-shadow: 0 0 0 1px rgba(0,0,0,.1);
```
Comment (talyssonoc) [inline app/assets/stylesheets/_color-picker.scss]: `rgba($black, .1);`

File: app/assets/stylesheets/_tag-group.scss [hunk 28]
```diff
@@ -0,0 +1,22 @@
+.tag-groups {
+  margin: 20px 0;
+
+  .color_container {
+    padding: 2px;
+    background-color: $white;
+    border-radius: 1px;
+    box-shadow: rgba(0, 0, 0, 0.0980392) 0 0 0 1px;
```
Comment (talyssonoc) [inline app/assets/stylesheets/_tag-group.scss]: `rgba($black, .0980392)`

**Comment by talyssonoc** [comment]
@brunohkbx could you add at least 1 functional/Capybara test for this feature?

File: spec/models/tag_group_spec.rb [hunk 29]
```diff
@@ -0,0 +1,41 @@
+require 'rails_helper'
+
+describe TagGroup, type: :model do
+
+  describe "#bg_color" do
+
+    context "with valid json" do
+      let(:valid_json) { "{\"r\":\"241\",\"g\":\"112\",\"b\":\"19\",\"a\":\"1\"}" }
```
Comment (andrerpbts) [inline spec/models/tag_group_spec.rb]: Kinda hard to read this...

If you use single quote, the backslashes will not be necessary anymore. Or, you could do
```ruby
 let(:valid_json) { { r: 241, g: 112, b: 19, a: 1 }.to_json }
```

**Comment by brunohkbx** [comment]
@talyssonoc I added a test.
Can you check please?

**Comment by brunohkbx** [comment]
Could someone check the tests that I created for tag group policy?

File: spec/policies/tag_group_policy_spec.rb [hunk 30]
```diff
@@ -0,0 +1,32 @@
+require 'rails_helper'
+
+describe ProjectPolicy do
```
Comment (andrerpbts) [inline spec/policies/tag_group_policy_spec.rb]: TagGroupPolicy

File: spec/policies/tag_group_policy_spec.rb [hunk 31]
```diff
@@ -0,0 +1,32 @@
...
+  let(:project) { create :project }
+  let(:pundit_context) { PunditContext.new(current_team, current_user, current_project: project) }
+  let(:current_team) { current_user.teams.first }
+  let(:policy_scope) { TagGroupPolicy::Scope.new(pundit_context, Project).resolve.all }
+  subject { TagGroupPolicy.new(pundit_context, project) }
```
Comment (andrerpbts) [inline spec/policies/tag_group_policy_spec.rb]: described_class

File: spec/policies/tag_group_policy_spec.rb [hunk 32]
```diff
@@ -0,0 +1,32 @@
+require 'rails_helper'
+
+describe ProjectPolicy do
+  let(:project) { create :project }
+  let(:pundit_context) { PunditContext.new(current_team, current_user, current_project: project) }
+  let(:current_team) { current_user.teams.first }
+  let(:policy_scope) { TagGroupPolicy::Scope.new(pundit_context, Project).resolve.all }
```
Comment (andrerpbts) [inline spec/policies/tag_group_policy_spec.rb]: described_class

File: spec/policies/tag_group_policy_spec.rb [hunk 33]
```diff
@@ -0,0 +1,32 @@
...
+    before do
+      project.users << current_user
+      current_team.ownerships.create(project: project, is_owner: true)
+    end
+
+    context "for an admin" do
+      let(:current_user) { create :user, :with_team_and_is_admin }
+
+      %i[index show create new update edit].each do |action|
+        it { should permit(action) }
```
Comment (andrerpbts) [inline spec/policies/tag_group_policy_spec.rb]: it { is_expected.to permit(action) }

File: spec/policies/tag_group_policy_spec.rb [hunk 34]
```diff
@@ -0,0 +1,32 @@
...
+      end
+    end
+
+    context "for a user" do
+      let(:current_user) { create :user, :with_team }
+
+      %i[index show create new update edit].each do |action|
+        it { should_not permit(action) }
```
Comment (andrerpbts) [inline spec/policies/tag_group_policy_spec.rb]: it { is_expected.not_to permit(action) }

File: app/assets/javascripts/components/tag_groups/ColorPick.js [hunk 35]
```diff
@@ -0,0 +1,49 @@
...
+export default class ColorPick extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      displayColorPicker: false,
+      color: this.props.color
+    };
+
+    this.handleClick = () => {
+      this.setState({ displayColorPicker: !this.state.displayColorPicker });
```
Comment (talyssonoc) [inline app/assets/javascripts/components/tag_groups/ColorPick.js]: I think it's better to declare it in the class itself and just do the bind here. Like:

```js
constructor(props) {
  // ...
  this.handleClick = this.handleClick.bind(this);
  // ...
}

handleClick() {
  this.setState( ... );
}
```

File: app/assets/javascripts/components/tag_groups/ColorPick.js [hunk 36]
```diff
@@ -0,0 +1,49 @@
...
+        <div style={ styles.swatch } onClick={ this.handleClick } className="swatch">
+          <div style={ styles.color } className="color-box" />
+          <input type="hidden" value={ this.state.color } name="tag_group[bg_color]" />
+        </div>
+        { this.state.displayColorPicker && <div className="color-box-popover">
+          <div className="color-box-cover" onClick={ this.handleClose }/>
+          <TwitterPicker color={ this.state.color } onChange={ this.handleChange } />
+        </div> }
+      </div>
+    )
```
Comment (talyssonoc) [inline app/assets/javascripts/components/tag_groups/ColorPick.js]: It's missing a `;` here.

File: app/assets/javascripts/controllers/tag_groups/controller.js [hunk 37]
```diff
@@ -0,0 +1,55 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import ColorPick from 'components/tag_groups/ColorPick';
```
Comment (talyssonoc) [inline app/assets/javascripts/controllers/tag_groups/controller.js]: I think it's better to call it `ColorPicker`.

File: app/assets/javascripts/controllers/tag_groups/controller.js [hunk 38]
```diff
@@ -0,0 +1,55 @@
...
+
+  $(document).ready(function(){
+    $(document).bind('ajaxError', function(event, jqxhr, settings, exception){
+      $("#new_tag_group").render_form_errors( $.parseJSON(jqxhr.responseText) );
+    });
+  });
+
+  (function($) {
+
+    $.fn.modal_success = function(){
```
Comment (talyssonoc) [inline app/assets/javascripts/controllers/tag_groups/controller.js]: Why do we need that and why is it using snake_case?

File: app/assets/javascripts/controllers/tag_groups/controller.js [hunk 39]
```diff
@@ -0,0 +1,55 @@
...
+        let $input = $('input[name="' + model + '[' + field + ']"]');
+        $input
+          .closest('.form-group')
+          .addClass('has-error').find('.help-block').html( messages.join(' & ') );
+      });
+
+    };
+
+    $.fn.clear_previous_errors = function(){
+      $('.form-group.has-error', this).each(function(){
```
Comment (talyssonoc) [inline app/assets/javascripts/controllers/tag_groups/controller.js]: We should avoid using CSS classes for selecting elements, CSS classes are meant to be used for styles. Add data attributes for that.

File: app/assets/javascripts/controllers/tag_groups/controller.js [hunk 40]
```diff
@@ -0,0 +1,55 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import ColorPick from 'components/tag_groups/ColorPick';
+
+export default () => {
+  const bgColor = $('[name=bg-color]').data('current-color') || "#F17013";
+
+  ReactDOM.render(<ColorPick color ={ bgColor }/>, document.getElementById('tag_group_color'));
+
+  $(document).ready(function(){
```
Comment (talyssonoc) [inline app/assets/javascripts/controllers/tag_groups/controller.js]: The controllers are already run inside a `$()` (https://github.com/Codeminer42/cm42-central/blob/master/app/assets/javascripts/app.js#L13) (that is an alias to `$(document).ready()`) so there's no need to do that.

File: app/assets/javascripts/controllers/tag_groups/controller.js [hunk 41]
```diff
@@ -0,0 +1,55 @@
...
+    $(document).bind('ajaxError', function(event, jqxhr, settings, exception){
+      $("#new_tag_group").render_form_errors( $.parseJSON(jqxhr.responseText) );
```
Comment (talyssonoc) [inline app/assets/javascripts/controllers/tag_groups/controller.js]: Let's avoid adding new jQuery methods.

File: app/assets/javascripts/controllers/tag_groups/controller.js [hunk 42]
```diff
@@ -0,0 +1,55 @@
...
+import ReactDOM from 'react-dom';
+import ColorPick from 'components/tag_groups/ColorPick';
+
+export default () => {
+  const bgColor = $('[name=bg-color]').data('current-color') || "#F17013";
+
+  ReactDOM.render(<ColorPick color ={ bgColor }/>, document.getElementById('tag_group_color'));
+
+  $(document).ready(function(){
+    $(document).bind('ajaxError', function(event, jqxhr, settings, exception){
```
Comment (talyssonoc) [inline app/assets/javascripts/controllers/tag_groups/controller.js]: Can't we simply add a listener for the form submit event?

File: app/controllers/tag_groups_controller.rb [hunk 43]
```diff
@@ -0,0 +1,55 @@
...
+    authorize @tag_group
+
+   respond_to do |format|
+      if @tag_group.save
+        format.html { redirect_to tag_groups_path }
+        format.js   { render action: 'show' }
+      else
+        format.html { redirect_to new_tag_group_path, flash: { error:@tag_group.errors.full_messages.to_sentence } }
+        format.json { render json: @tag_group.errors, status: :unprocessable_entity }
+        # added:
```
Comment (talyssonoc) [inline app/controllers/tag_groups_controller.rb]: Why is this comment needed?

File: app/views/tag_groups/_form.html.erb [hunk 44]
```diff
@@ -0,0 +1,26 @@
+<% modal ||= false %>
```
Comment (talyssonoc) [inline app/views/tag_groups/_form.html.erb]: Why is this needed?

File: app/views/tag_groups/_form.html.erb [hunk 45]
```diff
@@ -0,0 +1,26 @@
...
+<% remote = modal ? true : false %>
+
+<%= form_for(tag_group, remote: remote, html: {role: :form, 'data-model' => 'tag_group'}) do |f| %>
```
Comment (talyssonoc) [inline app/views/tag_groups/_form.html.erb]: You can avoid using strings for HTML attributes doing like this:

`{ role: :form, data: { model: 'tag_group' } }`

File: app/views/tag_groups/index.html.erb [hunk 46]
```diff
@@ -0,0 +1,45 @@
...
+                    <div class="color_container">
+                      <div class="color" style="background-color: <%= tag_group.bg_color %>;">
+                      </div>
+                    </div>
+                  </td>
+                  <td><%= tag_group.name %></td>
+                  <td><%= tag_group.description %></td>
+                  <td>
+                    <%= link_to t('edit'), edit_tag_group_path(tag_group) %>
+                    <%= link_to t('delete'), tag_group, method: :delete, class: 'danger', :data => {confirm: t('confirm')} %>
```
Comment (talyssonoc) [inline app/views/tag_groups/index.html.erb]: Prefer to use new hash syntax here instead of hash rocket.

File: config/locales/pt-BR.yml [hunk 47]
```diff
@@ -299,6 +300,13 @@ pt-BR:
     no_projects_found: "Oops! Parece que esta equipe não tem nenhum projeto."
     no_recent_activity: "Oops! Essa equipe não possui atividade recent."
 
+  tag_group:
+    color: "Cor"
+    name: "Nome"
+    description: "Descrição"
+    actions: "Ações"
+    bg_color: "Cor do Background"
```
Comment (talyssonoc) [inline config/locales/pt-BR.yml]: Maybe  "Cor do fundo" or "Cor de fundo" here?

File: config/locales/pt-BR.yml [hunk 48]
```diff
@@ -299,6 +300,13 @@ pt-BR:
...
+    tag_groups: "Tag Groups"
```
Comment (talyssonoc) [inline config/locales/pt-BR.yml]: Shouldn't this be in portuguese?

File: app/assets/javascripts/app.js [hunk 49]
```diff
@@ -1,10 +1,14 @@
 import './libs';
 import ProjectsIndexController from './controllers/projects/IndexController';
+import TagGroupsController from './controllers/tag_groups/controller';
```
Comment (talyssonoc) [inline app/assets/javascripts/app.js]: This file should have a better name like `TagGroupsController`, not just `Controller`.

File: config/locales/pt-BR.yml (hunk 48 above)
Comment (brunohkbx) [inline config/locales/pt-BR.yml]: What description do you suggest?

Comment (talyssonoc) [inline config/locales/pt-BR.yml]: What about "Grupos de rótulos"?

File: app/views/tag_groups/_form.html.erb [hunk 50]
```diff
@@ -0,0 +1,26 @@
...
+  </div>
+
+  <div class="form-group">
+    <%= f.label t('tag_group.description'), class: 'control-label' %>
+    <%= f.text_area :description, class: 'form-control auth-form-control' %>
+  </div>
+
+  <div class="form-group">
+    <%= f.label t('tag_group.bg_color'), class: 'control-label' %>
+    <div id='tag_group_color'></div>
```
Comment (eddiecode) [inline app/views/tag_groups/_form.html.erb]: Use double quotes here.

File: app/views/tag_groups/_form.html.erb (hunk 44 above)
Comment (brunohkbx) [inline app/views/tag_groups/_form.html.erb]: This conditional is responsible to render the form with remote true or false.
It will be true when the form is rendered inside the modal at this line : https://github.com/Codeminer42/cm42-central/blob/bfcb5ae86fbcd0365a103a47b2ddcdce76113ba0/app/views/projects/_form.html.erb#L81

Comment (talyssonoc) [inline app/views/tag_groups/_form.html.erb]: And when will this not be passed to justify the use of `||=`? If you don't pass a local variable to a partial it'll simply break so this `||=` here won't work.

Comment (brunohkbx) [inline app/views/tag_groups/_form.html.erb]: It will not be passed in the actions new and edit.
https://github.com/Codeminer42/cm42-central/blob/bfcb5ae86fbcd0365a103a47b2ddcdce76113ba0/app/views/tag_groups/new.html.erb#L17

Comment (talyssonoc) [inline app/views/tag_groups/_form.html.erb]: Prefer to pass it explicitly there.

File: app/assets/javascripts/controllers/tag_groups/controller.js (hunk 38 above)
Comment (brunohkbx) [inline app/assets/javascripts/controllers/tag_groups/controller.js]: This function is called every time the ajax request succed.
Check that:
https://github.com/Codeminer42/cm42-central/blob/cc88e955fc8c7231d0889fd25594f87e80add41f/app/views/tag_groups/show.js.erb#L2

PS : Changed to camelCase.

File: app/assets/javascripts/components/tag_groups/ColorPicker.js [hunk 51]
```diff
@@ -0,0 +1,55 @@
...
+  constructor(props) {
+    super(props);
+    this.state = {
+      displayColorPicker: false,
+      color: this.props.color
+    };
+
+    this.handleClick = this.handleClick.bind(this);
+
+    this.handleClose = this.handleClick.bind(this);
```
Comment (talyssonoc) [inline app/assets/javascripts/components/tag_groups/ColorPicker.js]: Shouldn't this be `handleClose.bind(this)` here?

File: app/assets/javascripts/components/tag_groups/ColorPicker.js [hunk 52]
```diff
@@ -0,0 +1,55 @@
...
+  handleClose(){
+    this.setState({ displayColorPicker: false });
+  };
+
+  handleChange(color){
+    this.setState({ color: color.hex });
+  };
+
+  render() {
+    const styles = reactCSS({
```
Comment (talyssonoc) [inline app/assets/javascripts/components/tag_groups/ColorPicker.js]: We should __not__ use it here, it's meant to use simple `style={{ backgroundColor: this.state.color }}` here on the tag we want this color. We shouldn't add a whole new dependency just because of a single dynamic style, I see no reason to add reactcss to the project right now.

File: app/assets/javascripts/controllers/tag_groups/TagGroupsController.js [hunk 53]
```diff
@@ -0,0 +1,48 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import ColorPicker from 'components/tag_groups/ColorPicker';
+
+export default () => {
+  const bgColor = $('[name=bg-color]').data('current-color') || "#F17013";
+
+  ReactDOM.render(<ColorPicker color ={ bgColor }/>, document.getElementById('tag_group_color'));
```
Comment (talyssonoc) [inline app/assets/javascripts/controllers/tag_groups/TagGroupsController.js]: Remove the space after the `color` prop here.

File: app/assets/javascripts/controllers/tag_groups/TagGroupsController.js [hunk 54]
```diff
@@ -0,0 +1,48 @@
...
+      $.each(errors, function(field, messages){
+        let $input = $('input[name="' + model + '[' + field + ']"]');
+        $input
+          .closest('.form-group')
+          .addClass('has-error').find('.help-block').html( messages.join(' & ') );
+      });
+
+    };
+
+    $.fn.clearPreviousE rrors = function(){
```
Comment (talyssonoc) [inline app/assets/javascripts/controllers/tag_groups/TagGroupsController.js]: There's an extra space here.

File: app/assets/stylesheets/_tag-group.scss [hunk 55]
```diff
@@ -0,0 +1,30 @@
...
+
+    .color {
+      width: 36px;
+      height: 14px;
+      border-radius: 2px;
+    }
+  }
+
+  .danger
+  {
```
Comment (talyssonoc) [inline app/assets/stylesheets/_tag-group.scss]: The `{` should be on the same line as `.danger`.

File: app/assets/javascripts/components/tag_groups/ColorPicker.js [hunk 56]
```diff
@@ -0,0 +1,55 @@
+import React from 'react';
+import reactCSS from 'reactcss';
```
Comment (talyssonoc) [inline app/assets/javascripts/components/tag_groups/ColorPicker.js]: We should not be using reactcss

File: app/views/projects/_form.html.erb [hunk 57]
```diff
@@ -54,3 +66,21 @@
...
+
+<!-- Modal -->
+<div class="modal fade" id="tag-group-modal" role="dialog">
+  <div class="modal-dialog">
+
+    <!-- Modal content-->
+    <div class="modal-content">
+      <div class="modal-header">
+        <button type="button" class="close" data-dismiss="modal">&times;</button>
+        <h4 class="modal-title">Create new tag group</h4>
```
Comment (eddiecode) [inline app/views/projects/_form.html.erb]: You forgot the translation here.

File: app/assets/javascripts/components/tag_groups/ColorPicker.js (hunk 51 above)
Comment (brunohkbx) [inline app/assets/javascripts/components/tag_groups/ColorPicker.js]: Without the 'this' keyword here I got this error : 
Uncaught ReferenceError: handleClick is not defined

Comment (talyssonoc) [inline app/assets/javascripts/components/tag_groups/ColorPicker.js]: The problem isn't the `this`, I omitted it just to be quicker, the problem is that the method to be bounded here is `handleClose`, not `handleClick`.

File: app/assets/javascripts/controllers/tag_groups/TagGroupsController.js [hunk 58]
```diff
@@ -0,0 +1,48 @@
...
+
+    };
+
+    $.fn.clearPreviousErrors = function(){
+      $('.form-group.has-error', this).each(function(){
+        $('.help-block', $(this)).html('');
+        $(this).removeClass('has-error');
+      });
+    }
+  }(jQuery));
```
Comment (talyssonoc) [inline app/assets/javascripts/controllers/tag_groups/TagGroupsController.js]: You don't need to do that, we have an alias to `$` pointing to jQuery already: https://github.com/Codeminer42/cm42-central/blob/master/config/webpack.config.js#L80
